### PR TITLE
CORDA-3018 Whitelisting attachments by public key - phase two tooling

### PR DESCRIPTION
**Pull request by lankydan**
Allow node operators to blacklist signing keys. These blacklisted keys prevent attachments that are received over the network from being _trusted_. The docs have been updated to detail how to generate the key hashes that the config requires.

A new shell command `attachments trustRoots` has been added to see what attachments exist on the node along with information about their trust and where it comes from.

`run dumpCheckpoints` has been replaced by `checkpoints dump` as `InternalCordaRPCOps` needed to change to prevent a function that is meant to be internal from being visible on the shell.

An example of the output can be seen below: 
attachment trust roots screen shots

### CODE REVIEW

File: docs/source/checkpoint-tooling.rst [hunk 1]
```diff
@@ -339,14 +339,14 @@ Note that "In progress" indicates the flows above have not completed (and will h
 
 .. note:: Always search for the flow id, in this case **90613d6f-be78-41bd-98e1-33a756c28808**
 
-2. From the CRaSH shell run the ``dumpCheckpoints`` command to trigger diagnostics information.
+2. From the CRaSH shell run the ``checkpoints dump`` command to trigger diagnostics information.
```
Comment (JamesHR3) [inline docs/source/checkpoint-tooling.rst]: A quick search suggests you're missing a couple of these in `changelog` and `upgrading-cordapps`

File: docs/source/corda-configuration-file.rst [hunk 2]
```diff
@@ -69,6 +69,14 @@ attachmentCacheBound
 
   *Default:* 1024
 
+blacklistedAttachmentSigningKeys
+  List of public keys (SHA-256 of public key) that prevents the node from accepting attachments over the network that are signed by any of these keys.
```
Comment (JamesHR3) [inline docs/source/corda-configuration-file.rst]: I'd rephrase to "List of SHA-256 hashes of public keys. Attachments signed by any of these public keys will not be considered as trust roots for any attachments received over the network."

File: testing/test-utils/src/main/kotlin/net/corda/testing/services/MockAttachmentStorage.kt [hunk 3]
```diff
@@ -22,9 +22,11 @@ import java.util.jar.Attributes
 import java.util.jar.JarInputStream
 
 /**
- * A mock implementation of [AttachmentStorage] for use within tests
+ * A mock implementation of [AttachmentStorageInternal] for use within tests
  */
-class MockAttachmentStorage : AttachmentStorage, SingletonSerializeAsToken() {
+class MockAttachmentStorage(override val blacklistedAttachmentSigningKeys: List<SecureHash> = emptyList()) :
+    AttachmentStorageInternal, SingletonSerializeAsToken() {
```
Comment (JamesHR3) [inline testing/test-utils/src/main/kotlin/net/corda/testing/services/MockAttachmentStorage.kt]: Annoyingly, I don't think you can implement `AttachmentStorageInternal` on this, because `MockAttachmentsStorage` is part of the public API (albeit the testing API). Doing so will expose internal functions externally.

File: node/src/main/kotlin/net/corda/node/internal/cordapp/CordappProviderImpl.kt [hunk 4]
```diff
@@ -70,20 +70,28 @@ open class CordappProviderImpl(val cordappLoader: CordappLoader,
     fun getCordappAttachmentId(cordapp: Cordapp): SecureHash? = cordappAttachments.inverse()[cordapp.jarPath]
 
     private fun loadContractsIntoAttachmentStore(): Map<SecureHash, URL> =
-            cordapps.filter { !it.contractClassNames.isEmpty() }.map {
-                it.jarPath.openStream().use { stream ->
+            cordapps.filter { it.contractClassNames.isNotEmpty() }.map { cordapp ->
+                cordapp.jarPath.openStream().use { stream ->
                     try {
                         // We can't make attachmentStorage a AttachmentStorageInternal as that ends up requiring
                         // MockAttachmentStorage to implement it.
```
Comment (JamesHR3) [inline node/src/main/kotlin/net/corda/node/internal/cordapp/CordappProviderImpl.kt]: I know this isn't directly related to this PR, but this comment is missing a bit of context that would probably be helpful to the next person editing stuff in this area. It should say that `MockAttachmentStorage` cannot implement `AttachmentStorageInternal` because doing so results in internal functions being exposed in the public API.

File: node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt [hunk 5]
```diff
@@ -227,18 +237,38 @@ class NodeAttachmentService(
             }
     }
 
-    private class AttachmentImpl(override val id: SecureHash, dataLoader: () -> ByteArray, private val checkOnLoad: Boolean, uploader: String?) : AbstractAttachment(dataLoader, uploader), SerializeAsToken {
+    @VisibleForTesting
```
Comment (JamesHR3) [inline node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt]: I think we're better off leaving this private and in the corresponding test not checking for attachment classes (I'll leave another comment pointing out where I mean).

File: node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt [hunk 6]
```diff
@@ -227,18 +237,38 @@ class NodeAttachmentService(
...
+
         override fun open(): InputStream {
             val stream = super.open()
             // This is just an optional safety check. If it slows things down too much it can be disabled.
             return if (checkOnLoad && id is SecureHash.SHA256) HashCheckingStream(id, attachmentData.size, stream) else stream
         }
 
-        private class Token(private val id: SecureHash, private val checkOnLoad: Boolean, private val uploader: String?) : SerializationToken {
-            override fun fromToken(context: SerializeAsTokenContext) = AttachmentImpl(id, context.attachmentDataLoader(id), checkOnLoad, uploader)
+        class Token(
```
Comment (JamesHR3) [inline node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt]: Does this need to be public?

File: node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt [hunk 7]
```diff
@@ -471,27 +525,30 @@ class NodeAttachmentService(
...
 
     private fun getContractAttachmentVersions(contractClassName: String): NavigableMap<Version, AttachmentIds> = contractsCache.get(contractClassName) { name ->
-        val attachmentQueryCriteria = AttachmentQueryCriteria.AttachmentsQueryCriteria(contractClassNamesCondition = Builder.equal(listOf(name)),
-                versionCondition = Builder.greaterThanOrEqual(0), uploaderCondition = Builder.`in`(TRUSTED_UPLOADERS))
-        val attachmentSort = AttachmentSort(listOf(AttachmentSort.AttachmentSortColumn(AttachmentSort.AttachmentSortAttribute.VERSION, Sort.Direction.DESC),
-                AttachmentSort.AttachmentSortColumn(AttachmentSort.AttachmentSortAttribute.INSERTION_DATE, Sort.Direction.DESC)))
+        val attachmentQueryCriteria = AttachmentQueryCriteria.AttachmentsQueryCriteria(
+            contractClassNamesCondition = Builder.equal(listOf(name)),
+            versionCondition = Builder.greaterThanOrEqual(0),
+            uploaderCondition = Builder.`in`(TRUSTED_UPLOADERS)
```
Comment (JamesHR3) [inline node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt]: This is interesting. I think this means if we receive a transaction with the same contract class as one we've already installed but at a later version, we'll be able to verify newer ones but will continue to use the older one. I think that's probably fine (an explicit install is required to use the new contract version), but thought it worth calling out.

File: node/src/test/kotlin/net/corda/node/services/persistence/NodeAttachmentServiceTest.kt [hunk 8]
```diff
@@ -1186,6 +1253,159 @@ class NodeAttachmentServiceTest {
...
+
+            val result = storage.getAllAttachments()
+            assertEquals(4, result.size)
+            assertThat(result.map { (_, attachment) -> attachment.id }).containsOnly(
+                attachmentA, attachmentB, attachmentC, attachmentD
+            )
+            assertThat(result.map { (name, _) -> name }).containsOnly(
+                "A.jar", "B.jar", "C.zip", null
+            )
+            assertThat(result.map { (_, attachment) -> attachment::class }).containsOnly(
```
Comment (JamesHR3) [inline node/src/test/kotlin/net/corda/node/services/persistence/NodeAttachmentServiceTest.kt]: Not sure this check is worth making, as the user of the API only knows that they've got some `Attachments` (i.e. it's not in the contract of the API).

File: tools/shell/src/main/kotlin/net/corda/tools/shell/AttachmentTrustTable.kt [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+import net.corda.core.internal.P2P_UPLOADER
+import org.crsh.text.Color
+import org.crsh.text.Decoration
+import org.crsh.text.RenderPrintWriter
+import org.crsh.text.ui.LabelElement
+import org.crsh.text.ui.Overflow
+import org.crsh.text.ui.RowElement
+import org.crsh.text.ui.TableElement
+
+class AttachmentTrustTable(
```
Comment (JamesHR3) [inline tools/shell/src/main/kotlin/net/corda/tools/shell/AttachmentTrustTable.kt]: I guess we didn't go for bolding the trust roots in the end? Similarly for displaying trusted keys?

File: core/src/main/kotlin/net/corda/core/internal/node/services/AttachmentStorageInternal.kt [hunk 10]
```diff
@@ -1,11 +1,15 @@
-package net.corda.node.services.persistence
+package net.corda.core.internal.node.services
```
[the remark this answers is not in the document]
Comment (lankydan) [reply core/src/main/kotlin/net/corda/core/internal/node/services/AttachmentStorageInternal.kt]: I mentioned this to @shamsasari before I made the change and he seemed ok with it. If `AttachmentStorageCoreInternal` is preferred I can always revert the move.

File: tools/shell/src/main/kotlin/net/corda/tools/shell/AttachmentTrustTable.kt (hunk 9 above)
Comment (lankydan) [reply tools/shell/src/main/kotlin/net/corda/tools/shell/AttachmentTrustTable.kt]: went for the `Installed` column instead. I will add a command to display trusted and untrusted keys + another for that displays all the signers of a particular attachment

File: core/src/main/kotlin/net/corda/core/internal/node/services/AttachmentStorageInternal.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (JamesHR3) [reply core/src/main/kotlin/net/corda/core/internal/node/services/AttachmentStorageInternal.kt]: I don't think we want to move this. The interface exposes methods that allows for uploading attachments as a trusted uploader. With the interface in node, cordapps know nothing about this at all. Once it's in core, it becomes possible for cordapps to cast to this interface and call the methods (it's not supported of course, but the possibility is there). That is potentially a security hole, so it's probably best not to move it.

On a separate note, it feels like this shouldn't need to be in core anyway, as I think the only thing you're adding to it is some node configuration?

[the remark this answers is not in the document]
Comment (shamsasari) [reply core/src/main/kotlin/net/corda/core/internal/node/services/AttachmentStorageInternal.kt]: Yes, after speaking to James it makes more sense to not move this internal interface because of the methods it has, even though it would be internal in core. Sorry about that.

File: node/src/main/kotlin/net/corda/node/services/attachments/NodeAttachmentTrustCalculator.kt [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+/**
+ * Implementation of [AttachmentTrustCalculator].
+ *
+ * @param keysToBlacklist Is a string representation of a SHA-256 [SecureHash] which is converted
+ * during initialisation of this class.
+ */
+class NodeAttachmentTrustCalculator(
+    private val attachmentStorage: AttachmentStorageInternal,
+    keysToBlacklist: List<String> = emptyList()
```
Comment (JamesHR3) [inline node/src/main/kotlin/net/corda/node/services/attachments/NodeAttachmentTrustCalculator.kt]: The fact that this is a list of strings is only because we're expecting to construct this using the node configuration. What this class really wants is a list of `SecureHash` - if it were constructed in another way for example, it would be a bit unfortunate to convert a list of hashes into strings only to convert them back again! I think we should move the configuration parsing bit of code out of this class and have it take `SecureHash` on the constructor.

File: node/src/main/kotlin/net/corda/node/services/attachments/NodeAttachmentTrustCalculator.kt [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+    /**
+     * Attachments signed by any of these public keys will not be considered as trust roots for any
+     * attachments received over the network.
+     *
+     * The list consists of SHA-256 hashes of public keys
+     */
+    private val blacklistedAttachmentSigningKeys: List<SecureHash> = keysToBlacklist.mapNotNull {
+        try {
+            SecureHash.parse(it)
+        } catch (e: IllegalArgumentException) {
```
Comment (JamesHR3) [inline node/src/main/kotlin/net/corda/node/services/attachments/NodeAttachmentTrustCalculator.kt]: Think we need to fail if these don't parse. It implies that someone has tried to blacklist something and got it wrong - it would be a serious issue if we then decided to go ahead anyway and allow whatever it is they were trying to blacklist through! It's also consistent with what `makeCordappLoader` does in `AbstractNode`.

File: core/src/main/kotlin/net/corda/core/transactions/WireTransaction.kt [hunk 13]
```diff
@@ -144,7 +144,13 @@ class WireTransaction(componentGroups: List<ComponentGroup>, val privacySalt: Pr
                 { null },
                 // Returning a dummy `missingAttachment` Attachment allows this deprecated method to work and it disables "contract version no downgrade rule" as a dummy Attachment returns version 1
                 { resolveAttachment(it.txhash) ?: missingAttachment },
-                { isAttachmentTrusted(it, null) }
+                {
+                    when (it) {
```
Comment (JamesHR3) [inline core/src/main/kotlin/net/corda/core/transactions/WireTransaction.kt]: This block of code is so common that I think it might be worth factoring it out somewhere. The current done thing seems to be to stick utility functions like this above `AbstractAttachment` (see `isUploaderTrusted`) - let's go with that for now.

File: node/src/main/kotlin/net/corda/node/services/attachments/NodeAttachmentTrustCalculator.kt [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+            }
+        }
+    }
+
+    override fun calculateAllTrustRoots(): List<AttachmentTrustRoot> {
+        val attachments = attachmentStorage.getAllAttachments()
+        return attachments.map { it.calculateAttachmentTrustRoot(attachments) }
+    }
+
+    private fun Pair<String?, Attachment>.calculateAttachmentTrustRoot(
```
Comment (JamesHR3) [inline node/src/main/kotlin/net/corda/node/services/attachments/NodeAttachmentTrustCalculator.kt]: Think this is O(n^2) in the number of attachments in the worst case as we iterate through every attachment, and for each attachment we potentially iterate through the list again if we need to look for a trust root. I think if we enhance the cache a bit we can do a bit better:
 - Do an attachment query for all attachments uploaded by trusted uploaders. These are the only things that can act as a trust root. Have the cache map keys to an object that says whether the key is trusted and if it is some details about the attachment that acts as the root. (I'd suggest a sealed class with an `Untrusted` and `Trusted` variant, and have the trusted variant hold the extra data).
 - Go through each attachment. If it's trusted by uploader then you've immediately got all the data you need. If it isn't, then check to see if each key is in the cache, and if you find it in there as `Trusted`, then again you've got the information you need.

Does that work?

File: node/src/main/kotlin/net/corda/node/services/persistence/AttachmentStorageInternal.kt [hunk 15]
```diff
@@ -16,4 +18,9 @@ interface AttachmentStorageInternal : AttachmentStorage {
      * Similar to above but returns existing [AttachmentId] instead of throwing [DuplicateAttachmentException]
      */
     fun privilegedImportOrGetAttachment(jar: InputStream, uploader: String, filename: String?): AttachmentId
-}
+
+    /**
+     * Get all attachments stored within the node paired to their file name's.
+     */
+    fun getAllAttachments(): List<Pair<String?, Attachment>>
```
Comment (JamesHR3) [inline node/src/main/kotlin/net/corda/node/services/persistence/AttachmentStorageInternal.kt]: Do we need to concern ourselves with the case where there's a lot of attachments? If our plan is to iterate over them anyway I think we're fine to use `query.stream` in this function, but I'm not sure this can return a list in that case. Whether this is necessary depends on how large we expect the number of attachments to be - might be worth checking if we get away with this.

Also, very minor point: name's -> names in the comment above.

File: node/src/test/kotlin/net/corda/node/services/attachments/AttachmentTrustCalculatorTest.kt [hunk 16]
```diff
@@ -0,0 +1,565 @@
...
+    private lateinit var database: CordaPersistence
+    private lateinit var storage: NodeAttachmentService
+    private lateinit var attachmentTrustCalculator: AttachmentTrustCalculator
+    private val services = rigorousMock<ServicesForResolution>().also {
+        doReturn(testNetworkParameters()).whenever(it).networkParameters
+    }
+
+    @Before
+    fun setUp() {
+        LogHelper.setLevel(PersistentUniquenessProvider::class)
```
Comment (JamesHR3) [inline node/src/test/kotlin/net/corda/node/services/attachments/AttachmentTrustCalculatorTest.kt]: Do we need this for anything?

**Comment by gendal** [comment]
Sorry for the drive-by... I'm just browsing through PRs before my return from holiday and this one caught my eye.  Is there an overall design for this anywhere? The linked Jira has a comment about a design being forthcoming from @JamesHR3 but I can't see one.

The thoughts that went through my mind were (without reading the whole thing so sorry if this is addressed and I just missed it):

* Do we need to worry about situations where a key is blacklisted but the contract Jar itself is signed by a surplus of signatures so that, if the blacklisted signature were not present, it would be the attachment would still trusted by the node? If so, which side do we come down on? Still reject the attachment? Or allow?
* Have we completely satisfied ourselves this works with upgrades? I can't tell if we're creating new database tables/columns or not from a quick glance
* We're OK with situations where there are attachments already loaded/processed and we then blacklist a key?

Sorry for what could be seen as lazy questions from a drive-by!

**Comment by lankydan** [comment]
@gendal the design doc is here - https://github.com/corda/enterprise/pull/2191/files#diff-3cf99577c794e5ebc3a66fbfe9995cee (put together by @JamesHR3)

This pr is for the second phase design.

Currently, if an attachment contains a blacklisted key, it will be rejected from loading into the contract sandbox (it will still be saved to the vault), and therefore will fail any transactions initiated with it. Even if it is signed by other trusted signers, it will still be rejected. If the blacklisted key is removed, there will be nothing stopping the once untrusted/rejected attachment from being accepted in the future.

No new tables or columns are being created.

I need to think about the last point more, and probably add a test case for it. Hopefully, the following sentence from the design adds a bit of clarity:

>  the node will not execute code received from the network if one of its signing keys is compromised, while still allowing historic transactions to be verified (assuming the original attachments are uploaded directly by the node operator).

File: node/src/main/kotlin/net/corda/node/services/persistence/AttachmentStorageInternal.kt [hunk 17]
```diff
@@ -16,4 +20,10 @@ interface AttachmentStorageInternal : AttachmentStorage {
      * Similar to above but returns existing [AttachmentId] instead of throwing [DuplicateAttachmentException]
      */
     fun privilegedImportOrGetAttachment(jar: InputStream, uploader: String, filename: String?): AttachmentId
-}
+
+    /**
+     * Get all attachments as a [Stream], filtered by the input [AttachmentQueryCriteria],
+     * stored within the node paired to their file name's.
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/persistence/AttachmentStorageInternal.kt]: filenames

File: node/src/main/kotlin/net/corda/node/services/persistence/AttachmentStorageInternal.kt [hunk 18]
```diff
@@ -16,4 +20,10 @@ interface AttachmentStorageInternal : AttachmentStorage {
...
+     */
+    fun getAllAttachmentsByCriteria(criteria: AttachmentQueryCriteria = AttachmentQueryCriteria.AttachmentsQueryCriteria()): Stream<Pair<String?, Attachment>>
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/persistence/AttachmentStorageInternal.kt]: I think having `Attachment` first makes more sense in the pair

File: node/src/main/kotlin/net/corda/node/services/persistence/AttachmentStorageInternal.kt [hunk 19]
```diff
@@ -16,4 +20,10 @@ interface AttachmentStorageInternal : AttachmentStorage {
      * Similar to above but returns existing [AttachmentId] instead of throwing [DuplicateAttachmentException]
      */
     fun privilegedImportOrGetAttachment(jar: InputStream, uploader: String, filename: String?): AttachmentId
-}
+
+    /**
+     * Get all attachments as a [Stream], filtered by the input [AttachmentQueryCriteria],
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/persistence/AttachmentStorageInternal.kt]: You need to add that the Stream must be closed once used, and you'll need to do that in NodeAttachmentTrustCalculator

File: node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt [hunk 20]
```diff
@@ -429,23 +462,32 @@ class NodeAttachmentService(
...
         return database.transaction {
-            val session = currentDBSession()
-            val criteriaBuilder = session.criteriaBuilder
-
-            val criteriaQuery = criteriaBuilder.createQuery(DBAttachment::class.java)
-            val root = criteriaQuery.from(DBAttachment::class.java)
+            createAttachmentsQuery(
+                criteria,
+                sorting
+            ).resultList.map { AttachmentId.parse(it.attId) }
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt]: Hmm, we're loading the entire `DBAttachment` but only need `attId` column. Can you raise a JIRA ticket to fix this - a query that just selects the `attId` (unless of course you want to fix this now!)

File: node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt [hunk 21]
```diff
@@ -471,27 +513,30 @@ class NodeAttachmentService(
...
-            // prepare query for execution
-            val query = session.createQuery(criteriaQuery)
-
-            // execution
-            TreeMap(query.resultList.groupBy { it.version }.map { makeAttachmentIds(it, name) }.toMap())
+            TreeMap(
+                createAttachmentsQuery(
+                    attachmentQueryCriteria,
+                    attachmentSort
+                ).resultList.groupBy { it.version }.map { makeAttachmentIds(it, name) }.toMap()
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt]: If you do `toMap(TreeMap())` then you can get rid of the outer `TreeMap`

File: tools/shell/src/main/kotlin/net/corda/tools/shell/AttachmentTrustTable.kt [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+    init {
+        content = createTable()
+        createRows()
+        writer.print(content)
+    }
+
+    private fun createTable(): TableElement {
+        val table = TableElement(2, 3, 1, 1, 3).overflow(Overflow.WRAP).rightCellPadding(3)
+        val header =
+            RowElement(true).add("Name", "Hash/Id", "Installed", "Trusted", "Trust Root").style(
```
Comment (shamsasari) [inline tools/shell/src/main/kotlin/net/corda/tools/shell/AttachmentTrustTable.kt]: Hash/Id -> Attachment ID

File: core-tests/src/test/kotlin/net/corda/coretests/transactions/AttachmentsClassLoaderTests.kt [hunk 23]
```diff
@@ -374,4 +390,72 @@ class AttachmentsClassLoaderTests {
...
+            cacheFactory,
+            listOf(keyPairA.public.hash)
+        )
+
+        val classJar = fakeAttachment(
+            "/com/example/something/UntrustedClass.class",
+            "Signed by someone trusted"
+        ).inputStream()
+        classJar.use {
+            storage.importContractAttachment(
```
[the remark this answers is not in the document]
Comment (lankydan) [reply core-tests/src/test/kotlin/net/corda/coretests/transactions/AttachmentsClassLoaderTests.kt]: Updating the test name as the trusted attachment is part of the test but that is not clear from the test name

File: node/src/main/kotlin/net/corda/node/services/attachments/NodeAttachmentTrustCalculator.kt [hunk 24]
```diff
@@ -0,0 +1,146 @@
...
+ */
+class NodeAttachmentTrustCalculator(
+    private val attachmentStorage: AttachmentStorageInternal,
+    private val database: CordaPersistence?,
+    cacheFactory: NamedCacheFactory,
+    private val blacklistedAttachmentSigningKeys: List<SecureHash> = emptyList()
+) : AttachmentTrustCalculator, SingletonSerializeAsToken() {
+
+    @VisibleForTesting
+    constructor(
```
[the remark this answers is not in the document]
Comment (lankydan) [reply node/src/main/kotlin/net/corda/node/services/attachments/NodeAttachmentTrustCalculator.kt]: i wanted to prevent the database from not being set by _accident_, maybe another solution to achieve what i want is to make the primary constructor private and only expose a constructor where database cannot be null and keep the `@VisibleForTesting` constructor as well

File: node/src/main/kotlin/net/corda/node/services/attachments/NodeAttachmentTrustCalculator.kt [hunk 25]
```diff
@@ -0,0 +1,146 @@
...
+    override fun calculateAllTrustRoots(): List<AttachmentTrustInfo> {
+
+        val publicKeyToTrustRootMap = mutableMapOf<PublicKey, TrustedAttachment>()
+        val attachmentTrustInfos = mutableListOf<AttachmentTrustInfo>()
+
+        require(database != null) {
+            // This should never be hit, except for tests that have not been setup correctly to test internal code
+            "CordaPersistence has not been set"
+        }
+        database!!.transaction {
```
[the remark this answers is not in the document]
Comment (lankydan) [reply node/src/main/kotlin/net/corda/node/services/attachments/NodeAttachmentTrustCalculator.kt]: changed to `checkNotNull`, but it turns out you cant remove the `!!`

File: node/src/main/kotlin/net/corda/node/services/attachments/NodeAttachmentTrustCalculator.kt [hunk 26]
```diff
@@ -0,0 +1,146 @@
...
+            for ((name, attachment) in getTrustedAttachments()) {
+                attachment.signerKeys.forEach {
```
[the remark this answers is not in the document]
Comment (lankydan) [reply node/src/main/kotlin/net/corda/node/services/attachments/NodeAttachmentTrustCalculator.kt]: I would assume that this command would be accessed infrequently and therefore would not be much of an issue. Adding the keys to the cache here was just an opportunity to add them outside of classloading attachments. If we don't like this, we can just remove it. If we do still want to cache like you suggested, then a `start` function could be added. It is most likely that the cache would _warm up_ quite quickly when the node is processing transactions.

File: node/src/main/kotlin/net/corda/node/services/attachments/NodeAttachmentTrustCalculator.kt [hunk 27]
```diff
@@ -0,0 +1,146 @@
...
+            // add signers to the cache as this is a fully trusted attachment
+            attachment.signerKeys
+                .filterNot { it.isBlacklisted() }
+                .forEach { trustedKeysCache.put(it, true) }
+            return true
+        }
+
+        if (attachment.isSignedByBlacklistedKeys()) return false
+
+        return attachment.signerKeys.any { signer ->
```
[the remark this answers is not in the document]
Comment (lankydan) [reply node/src/main/kotlin/net/corda/node/services/attachments/NodeAttachmentTrustCalculator.kt]: this is wrong, if an attachment is signed by a blacklisted key at all we need to return `false`. By adding it into the cache lambda, it will return `true` if it hits a valid key before a blacklisted key.

[the remark this answers is not in the document]
Comment (shamsasari) [reply node/src/main/kotlin/net/corda/node/services/attachments/NodeAttachmentTrustCalculator.kt]: yes, you're right. What you could do is return false from within the lamdba, but i guess it's not that big a deal.

File: node/src/test/kotlin/net/corda/node/services/attachments/AttachmentTrustCalculatorTest.kt [hunk 28]
```diff
@@ -0,0 +1,628 @@
...
+
+            // Sanity check.
+            assertEquals(key1, key2, "Different public keys used to sign jars")
+            assertTrue(attachmentTrustCalculator.calculate(storage.openAttachment(v1Id)!!), "Initial attachment $v1Id should be trusted")
+            assertTrue(attachmentTrustCalculator.calculate(storage.openAttachment(v2Id)!!), "Other attachment $v2Id should be trusted")
+        }
+    }
+
+    @Test
+    fun `all non contract jars not trusted if all are uploaded by non trusted uploaders`() {
```
Comment (shamsasari) [inline node/src/test/kotlin/net/corda/node/services/attachments/AttachmentTrustCalculatorTest.kt]: This doesn't make sense!

File: node/src/test/kotlin/net/corda/node/services/attachments/AttachmentTrustCalculatorTest.kt [hunk 29]
```diff
@@ -0,0 +1,628 @@
...
+
+            // Sanity check.
+            assertEquals(key1, key2, "Different public keys used to sign jars")
+            assertFalse(attachmentTrustCalculator.calculate(storage.openAttachment(v1Id)!!), "Initial contract $v1Id should not be trusted")
+            assertFalse(attachmentTrustCalculator.calculate(storage.openAttachment(v2Id)!!), "Upgraded contract $v2Id should not be trusted")
+        }
+    }
+
+    @Test
+    fun `non contract jar trusted if trusted jar with same key present`() {
```
Comment (shamsasari) [inline node/src/test/kotlin/net/corda/node/services/attachments/AttachmentTrustCalculatorTest.kt]: use hyphen: non-contract

File: node/src/test/kotlin/net/corda/node/services/attachments/AttachmentTrustCalculatorTest.kt [hunk 30]
```diff
@@ -0,0 +1,628 @@
...
+
+            // Sanity check.
+            assertEquals(key1, key2, "Different public keys used to sign jars")
+            assertFalse(attachmentTrustCalculator.calculate(storage.openAttachment(v1Id)!!), "Initial attachment $v1Id should not be trusted")
+            assertFalse(attachmentTrustCalculator.calculate(storage.openAttachment(v2Id)!!), "Other attachment $v2Id should not be trusted")
+        }
+    }
+
+    @Test
+    fun `non contract jars not trusted if unsigned`() {
```
Comment (shamsasari) [inline node/src/test/kotlin/net/corda/node/services/attachments/AttachmentTrustCalculatorTest.kt]: non-contract

File: node/src/test/kotlin/net/corda/node/services/attachments/AttachmentTrustCalculatorTest.kt [hunk 31]
```diff
@@ -0,0 +1,628 @@
...
+    }
+
+    @Test
+    fun `calculateAllTrustRoots returns all attachment trust roots`() {
+        SelfCleaningDir().use { file ->
+            val aliasA = "dan"
+            val aliasB = "james"
+            val aliasC = "tudor"
+            val aliasD = "shams"
+            val password = "one day the attachment service will be refactored"
```
Comment (shamsasari) [inline node/src/test/kotlin/net/corda/node/services/attachments/AttachmentTrustCalculatorTest.kt]: :-D

File: node/src/test/kotlin/net/corda/node/services/attachments/AttachmentTrustCalculatorTest.kt [hunk 32]
```diff
@@ -0,0 +1,628 @@
...
+
+            val (zipC, _) = makeTestJar(listOf(Pair("file", "content")))
+            val (zipD, _) = makeTestJar(listOf(Pair("magic_file", "magic_content_puff")))
+
+            val attachmentA = jarSignedByA.read { storage.privilegedImportAttachment(it, "app", "A.jar") }
+            val attachmentB = jarSignedByAB.read { storage.privilegedImportAttachment(it, "untrusted", "B.jar") }
+            val attachmentC = zipC.read { storage.privilegedImportAttachment(it, "app", "C.zip") }
+            zipD.read { storage.privilegedImportAttachment(it, "untrusted", null) }
+
+            assertThat(attachmentTrustCalculator.calculateAllTrustRoots()).containsOnly(
```
Comment (shamsasari) [inline node/src/test/kotlin/net/corda/node/services/attachments/AttachmentTrustCalculatorTest.kt]: I just realised that the name `calculateAllTrustRoots()` is slightly incorrect. It's actually returning attachement trust information. So `calculateAttachmentsTrustInfo`?

File: node/src/test/kotlin/net/corda/node/services/attachments/AttachmentTrustCalculatorTest.kt [hunk 33]
```diff
@@ -0,0 +1,628 @@
...
+                    fileName = "A.jar",
+                    uploader = "app",
+                    trustRootId = attachmentA,
+                    trustRootFileName = "A.jar"
+                ),
+                AttachmentTrustInfo(
+                    attachmentId = attachmentB,
+                    fileName = "B.jar",
+                    uploader = "untrusted",
+                    trustRootId = null,
```
Comment (shamsasari) [inline node/src/test/kotlin/net/corda/node/services/attachments/AttachmentTrustCalculatorTest.kt]: I just realised that the name `calculateAllTrustRoots()` is slightly incorrect if it can include an element which doesn't have a root trust. Perhaps `calculateAllTrustInfo`?

File: node/src/test/kotlin/net/corda/node/services/persistence/NodeAttachmentServiceTest.kt [hunk 34]
```diff
@@ -1188,6 +904,88 @@ class NodeAttachmentServiceTest {
...
+                listOf(key), emptyList(), emptyList(), emptyList()
+            )
+        }
+    }
+
+    @Test
+    fun `getAllAttachmentsByCriteria returns attachments filtered by criteria`() {
+        SelfCleaningDir().use { file ->
+            val aliasA = "Dan"
+            val password = "i am so tired with this work"
```
Comment (shamsasari) [inline node/src/test/kotlin/net/corda/node/services/persistence/NodeAttachmentServiceTest.kt]: Don't give up now, you've done so well.

File: tools/shell/src/main/java/net/corda/tools/shell/AttachmentShellCommand.java [hunk 35]
```diff
@@ -0,0 +1,15 @@
...
+
+import org.crsh.cli.Command;
+import org.crsh.cli.Man;
+
+import static net.corda.tools.shell.InteractiveShell.runAttachmentTrustRootsView;
+
+public class AttachmentShellCommand extends InteractiveShellCommand {
+
+    @Command
+    @Man("Displays the trusted CorDapp attachments that have been manually installed or received over the network")
```
Comment (shamsasari) [inline tools/shell/src/main/java/net/corda/tools/shell/AttachmentShellCommand.java]: Or maybe that method should just be called `calculateTrustedAttachments`?

File: node/src/test/kotlin/net/corda/node/services/attachments/AttachmentTrustCalculatorTest.kt (hunk 32 above)
Comment (lankydan) [reply node/src/test/kotlin/net/corda/node/services/attachments/AttachmentTrustCalculatorTest.kt]: sure, i debated on whether to change the name or not

File: docs/source/node-administration.rst [hunk 36]
```diff
@@ -279,3 +279,45 @@ Here is a sample output displayed by the ``run nodeDiagnosticInfo`` command exec
...
+
+        - Whether an attachment is installed locally
+        - If an attachment is trusted
+        - Which other attachment, if any, provided trust to an attachment
+
+    Below is an example out the command's output:
+
+    .. code-block:: none
+
+        Name                                          Attachment ID                                                        Installed             Trusted                Trust Root
```
[the remark this answers is not in the document]
Comment (lankydan) [reply docs/source/node-administration.rst]: Added extra explanation
